src/GatewayAPI/Services/CoreNodesSupervisorService.cs

```diff
@@ -0,0 +1,209 @@
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Common.Extensions;
+using GatewayAPI.Configuration;
+using GatewayAPI.Configuration.Models;
+using GatewayAPI.CoreCommunications;
+using CoreApiModel = RadixCoreApi.Generated.Model;
+
+namespace GatewayAPI.Services;
+
+public interface ICoreNodesSupervisorService
+{
+    public CoreApiNode GetRandomHealthiestCoreNode();
+
+    public Task ReviseCoreNodesHealth(CancellationToken cancellationToken);
+}
+
+public class CoreNodesSupervisorService : ICoreNodesSupervisorService
+{
+    // Using explicit integers for enum values
+    // because they're used for ordering the nodes (from best to worst).
+    private enum CoreNodeStatus
+    {
+        HealthyAndSynced = 0,
+        Unknown = 1,
+        HealthyButLagging = 2,
+        Unhealthy = 3,
+    }
+
+    private static readonly IEnumerable<CoreNodeStatus> _statusesFromBestToWorst =
+        Enum.GetValues(typeof(CoreNodeStatus)).Cast<CoreNodeStatus>().OrderBy(s => s);
+
+    private readonly ILogger _logger;
+    private readonly HttpClient _httpClient;
+    private readonly CoreApiModel.NetworkIdentifier _networkIdentifier;
+    private readonly long _maxAllowedStateVersionLagToBeConsideredSynced;
+
+    private readonly Dictionary<CoreApiNode, CoreNodeStatus> _coreNodesStatuses;
+    private readonly object _coreNodesStatusesLock = new();
+
+    public CoreNodesSupervisorService(
+        ILogger<CoreNodesSupervisorService> logger,
+        HttpClient httpClient,
+        IGatewayApiConfiguration configuration)
+    {
+        _logger = logger;
+        _httpClient = httpClient;
+
+        _networkIdentifier = new CoreApiModel.NetworkIdentifier(configuration.GetNetworkName());
+        _maxAllowedStateVersionLagToBeConsideredSynced = configuration
+            .GetCoreApiNodeHealth().MaxAllowedStateVersionLagToBeConsideredSynced;
+
+        _coreNodesStatuses = configuration.GetCoreNodes()
+            .Where(n => n.IsEnabled && !string.IsNullOrWhiteSpace(n.CoreApiAddress))
+            .ToDictionary(n => n, _ => CoreNodeStatus.Unknown);
+    }
+
+    public CoreApiNode GetRandomHealthiestCoreNode()
+    {
+        lock (_coreNodesStatusesLock)
+        {
+            foreach (var status in _statusesFromBestToWorst)
+            {
+                var coreNodesOfStatus = GetCoreNodesOfStatus(status);
+                if (coreNodesOfStatus.Any())
+                {
+                    return coreNodesOfStatus.GetRandomBy(n => (double)n.RequestWeighting);
+                }
+            }
+        }
+
+        throw new Exception("No core nodes available");
```

Comment:
We should use a more specific Exception type so that the API surface knows how to map it, and give a friendly-ish response message for people setting up their servers and testing for the first time. I would suggest something like:

```
InternalServerException.OfNoValidCoreApiNodesAvailable(...)
```

Which has a helpful message; similar to eg `InternalServerException.OfInvalidGatewayException`